src/main/java/games/strategy/triplea/ai/proAI/util/ProPurchaseUtils.java

```diff
@@ -116,14 +117,17 @@ public static boolean canUnitsBePlaced(final List<Unit> units, final PlayerID pl
     return purchaseOptions;
   }
 
-  public static ProPurchaseOption randomizePurchaseOption(final Map<ProPurchaseOption, Double> purchaseEfficiencies,
-      final String type) {
+  public static Optional<ProPurchaseOption> randomizePurchaseOption(
+      final Map<ProPurchaseOption, Double> purchaseEfficiencies, final String type) {
 
     ProLogger.trace("Select purchase option for " + type);
     double totalEfficiency = 0;
     for (final Double efficiency : purchaseEfficiencies.values()) {
       totalEfficiency += efficiency;
     }
+    if (totalEfficiency == 0) {
+      return Optional.empty();
+    }
```

Comment:
This is essentially the core of the change. Everything else is just updating all the references and clean up.